crates/kubelet/src/node.rs

```diff
@@ -166,6 +169,142 @@ pub async fn create_node<P: 'static + Provider + Sync + Send>(
+        field_selector: Some(node_selector),
+        ..Default::default()
+    };
+    let kube::api::ObjectList { items: pods, .. } = pod_client.list(&params).await?;
+
+    let lp = ListParams::default().fields(&format!("spec.nodeName={}", node_name));
+
+    // The delete call may return a "pending" response, we must watch for the actual delete event.
+    let mut stream = pod_client.watch(&lp, "0").await?.boxed();
+
+    warn!("Evicting {} pods.", pods.len());
+
+    for pod in pods {
+        let pod = Pod::new(pod);
+        if pod.is_daemonset() {
+            warn!("Skipping eviction of DaemonSet '{}'", pod.name());
+            continue;
+        }
+        if pod.is_static() {
+            let mut container_statuses = HashMap::new();
+            for container in pod.containers() {
+                container_statuses.insert(
+                    container.name.clone(),
+                    ContainerStatus::Terminated {
+                        timestamp: Utc::now(),
+                        message: "Evicted on node shutdown.".to_string(),
+                        failed: false,
+                    },
+                );
+            }
+            let status = Status {
+                message: Some("Evicted on node shutdown.".to_string()),
+                container_statuses,
+            };
+            pod.patch_status(client.clone(), status).await;
+            warn!("Marked static pod as terminated.");
+            continue;
+        }
+        match evict_pod(&client, pod.name(), pod.namespace(), &mut stream).await {
+            Ok(_) => (),
+            Err(e) => {
+                // Absorb the error and attempt to delete other pods with best effort.
+                error!("Error evicting pod: {:?}", e)
+            }
+        }
+    }
+    Ok(())
+}
+
+type PodStream = std::pin::Pin<
+    Box<
+        dyn futures::Stream<Item = Result<kube::api::WatchEvent<KubePod>, kube::error::Error>>
+            + Send,
+    >,
+>;
+
+async fn evict_pod(
+    client: &kube::Client,
+    name: &str,
+    namespace: &str,
+    stream: &mut PodStream,
+) -> anyhow::Result<()> {
+    let ns_client: Api<KubePod> = Api::namespaced(client.clone(), namespace);
+    warn!("Evicting namespace '{}' pod '{}'", namespace, name);
+    let params = Default::default();
+    let response = ns_client.delete(name, &params).await?;
+
+    if response.is_left() {
+        // TODO Timeout?
+        warn!("Waiting for pod '{}' eviction.", name);
+        while let Some(event) = stream.try_next().await? {
+            match event {
+                kube::api::WatchEvent::Deleted(s) => {
+                    let pod = Pod::new(s);
+                    if (name == pod.name()) & (namespace == pod.namespace()) {
+                        warn!("Pod '{}' evicted.", name);
+                        break;
+                    }
+                }
+                _ => (),
+            }
+        }
+    } else {
+        warn!("Pod '{}' evicted.", name);
```

Comment:
Should all these logs be `debug` not `warn`?

Reply:
I think `info` is the right level here

Reply:
ok, will do